packages/workflow/src/WorkflowDataProxy.ts

```diff
@@ -1018,8 +1042,19 @@ export class WorkflowDataProxy {
 										itemIndex = that.itemIndex;
 									}
 
-									const executionData = that.connectionInputData;
-									const input = executionData[itemIndex];
+									const pinnedData = getPinDataIfManualExecution(
+										that.workflow,
+										nodeName,
+										that.mode,
+									);
+									const executionData = that.connectionInputData.length
+										? that.connectionInputData
+										: pinnedData;
+									const input = executionData?.[itemIndex];
+
+									if (pinnedData && !that.connectionInputData.length) {
+										return input;
+									}
```

Comment:
Finding this part a bit hard to follow, I'll revisit this bit tomorrow with fresh eyes.